File: 3516-design-neighbor-sum-service/design-neighbor-sum-service.cpp
class neighborSum {
public:
    int n,m;
    vector<int> dia[4]={{-1,-1},{-1,1},{1,-1},{1,1}};
    vector<int> adj[4]={{-1,0},{1,0},{0,-1},{0,1}};
    unordered_map<int,pair<int,int>> map;
    neighborSum(vector<vector<int>>& grid) 
    {
        n=grid.size();
        m=grid[0].size();
        for(int i=0;i<n;i++)
        {
            for(int j=0;j<m;j++)
            {
                int adjSum=0;
                int diaSum=0;
                for(int k=0;k<4;k++)
                {
                    int adjRow=i+adj[k][0];
                    int adjCol=j+adj[k][1];
                    if(adjRow>=0 && adjRow<n && adjCol>=0 && adjCol<m)
                    adjSum+=grid[adjRow][adjCol];

                    int diaRow=i+dia[k][0];
                    int diaCol=j+dia[k][1];
                    if(diaRow>=0 && diaRow<n && diaCol>=0 && diaCol<m)
                    diaSum+=grid[diaRow][diaCol];
                }

                map[grid[i][j]]={adjSum,diaSum};
            }
        }
    }
    
    int adjacentSum(int value) 
    {
        return map[value].first;
    }
    
    int diagonalSum(int value) 
    {
        return map[value].second;
    }
};

/**
 * Your neighborSum object will be instantiated and called as such:
 * neighborSum* obj = new neighborSum(grid);
 * int param_1 = obj->adjacentSum(value);
 * int param_2 = obj->diagonalSum(value);
 */

File: 3516-design-neighbor-sum-service/design-neighbor-sum-service.java
class NeighborSum {

    int[][] dia, adj, grid;
    Map<Integer, int[]> map;
    public NeighborSum(int[][] grid) 
    {
        map = new HashMap<>();
        this.grid = grid;
        dia = new int[][]{{-1,-1}, {-1,1}, {1,-1}, {1,1}};
        adj = new int[][]{{-1,0}, {1,0}, {0,-1}, {0,1}};

        for(int i = 0; i < grid.length; i++)
        {
            for(int j = 0; j < grid[0].length; j++)
            map.put(grid[i][j], new int[]{i, j});
        }
    }
    
    public int adjacentSum(int value) 
    {
        int[] coordinates = map.get(value);
        int r = coordinates[0], c = coordinates[1];

        int sum = 0;
        for(int i = 0; i < 4; i++)
        {
            int row = r + adj[i][0];
            int col = c + adj[i][1];
            if(row >= 0 && row < grid.length && col >= 0 && col < grid[0].length)
            sum += grid[row][col];
        }

        return sum;
    }
    
    public int diagonalSum(int value) 
    {
        int[] coordinates = map.get(value);
        int r = coordinates[0], c = coordinates[1];

        int sum = 0;
        for(int i = 0; i < 4; i++)
        {
            int row = r + dia[i][0];
            int col = c + dia[i][1];
            if(row >= 0 && row < grid.length && col >= 0 && col < grid[0].length)
            sum += grid[row][col];
        }

        return sum;
    }
}

/**
 * Your NeighborSum object will be instantiated and called as such:
 * NeighborSum obj = new NeighborSum(grid);
 * int param_1 = obj.adjacentSum(value);
 * int param_2 = obj.diagonalSum(value);
 */